### PR TITLE
refactor(proxy): Move binary code into `api` crate

### LINKED ISSUE #1052: Make it easier to detect unused code
At the moment all modules in the `api` crate (`proxy/api/src/lib.rs`) are public but only a fraction of these APIs are used for the proxy binary in `proxy/api/src/main.rs`. This results in unused code not being detected properly.

As a solution I’d suggest moving all the code from `proxy/api/src/main.rs` into a module in the crate and only export that module to build the binary.

### PR DESCRIPTION
**Pull request by geigerzaehler**
To benefit from all linting we move all code from `api/src/main.rs` to `api/src/lib.rs`.

Fixes #1052

### CODE REVIEW

**Review by xla — CHANGES_REQUESTED** [review]
Conceptually this change is nice. There are a couple of changes that would put things in the right place.

* in a large crate like this the `lib.rs` should be purely `mod` declarations
* operational concerns still belong in `main.rs` translation of binary arguments, init of logging
* the rigging and run loop shenanigans might be best moved to something like `node` module
* the main in lib code should be thoroughly avoided as it is not a sensible API for anything but a binary

**Comment by xla** [comment]
As inspiration the work currently done for the seed node in radicle-bins can be used as inspiration.

**Comment by geigerzaehler** [comment]
> operational concerns still belong in `main.rs` translation of binary arguments, init of logging

What code specifically are you talking about? Code like `env::set_if_unset` comes from the crate. Using it in `main.rs` kind of defeats the purpose.

In addition, the `Args` struct is used, for example, by `rig` so we can’t just keep it in `main`. Let’s maybe try to not get caught up in getting the separation perfect.

> the rigging and run loop shenanigans might be best moved to something like `node` module

How do you feel about using a `process` module?

**Comment by xla** [comment]
> In addition, the `Args` struct is used, for example, by `rig` so we can’t just keep it in `main`. Let’s maybe try to not get caught up in getting the separation perfect.

Everything that translates from and to OS APIs, primitives into domain

* ENV variables mapping to config
* reading args mapping it to config
* exit codes
* STDIN/STDOU/STDERR

naturally belong in the main where the translation into the domain types are happening.

> How do you feel about using a `process` module?

Sure, not too fuzzed about the actual name. :>

File: proxy/api/src/project.rs [hunk 1]
```diff
@@ -140,15 +140,6 @@ impl Projects {
 
         Ok(projects)
     }
-
-    /// Give back an `Iter` that can be used to iterate over the projects. It first yields
-    /// contributed projects and then tracked projects.
-    pub fn iter(&self) -> Iter<'_> {
```
Comment (FintanH) [inline proxy/api/src/project.rs:146]: I mean, it's unused but doesn't mean it's not useful for the future :thinking: How come it doesn't trigger on the `Iter` struct too?

**Comment by geigerzaehler** [comment]
> Everything that translates from and to OS APIs, primitives into domain

Could you have another look? This should be addressed now.

**Review by FintanH — COMMENTED** [review]
Overall, I'm digging the reorganisation :+1:

File: proxy/api/src/project.rs (hunk 1 above)
Comment (geigerzaehler) [reply proxy/api/src/project.rs:146]: This is the same code as for `<Projects as IntoIter>::into_iter()` which probably is used implicitly through a `for .. of`.

Comment (FintanH) [reply proxy/api/src/project.rs:146]: Ah very good, thanks :grin:

**Review by FintanH — APPROVED** [review]

**Review by rudolfs — APPROVED** [review]

**Review by FintanH — APPROVED** [review]

**Review by xla — APPROVED** [review]
🏭 📱 ♉️ 🔉

**Pull request merged**